databricks_cli/pipelines/api.py

```diff
@@ -0,0 +1,182 @@
+        Local libraries are (currently) jar files with a file scheme or no scheme at all.
+        All other libraries should be present in a supported external source.
+        :param lib_objects: List[LibraryObject]
+        :return: List[List[LibraryObject], List[LibraryObject]] ([Local, External])
+        """
+        local_lib_objects, external_lib_objects = [], []
+        for lib_object in lib_objects:
+            parsed_uri = urllib.parse.urlparse(lib_object.path)
+            if lib_object.lib_type == 'jar' and parsed_uri.scheme == '':
+                local_lib_objects.append(lib_object)
+            elif lib_object.lib_type == 'jar' and parsed_uri.scheme.lower() == 'file':
+                # exactly 1 or 3
+                if parsed_uri.path.startswith('//') or parsed_uri.netloc != '':
+                    raise RuntimeError('invalid file uri scheme, '
+                                       'did you mean to use file:/ or file:///')
+                local_lib_objects.append(LibraryObject(lib_object.lib_type, parsed_uri.path))
+            else:
+                external_lib_objects.append(lib_object)
+        return local_lib_objects, external_lib_objects
+
+    def _upload_local_libraries(self, local_lib_objects):
+        remote_lib_objects = [LibraryObject(llo.lib_type, self._get_hashed_path(llo.path))
+                              for llo in local_lib_objects]
+
+        transformed_remote_lib_objects = [LibraryObject(rlo.lib_type, DbfsPath(rlo.path))
+                                          for rlo in remote_lib_objects]
+        upload_files = [llo_tuple for llo_tuple in
+                        zip(local_lib_objects, transformed_remote_lib_objects)
+                        if not self.dbfs_client.file_exists(llo_tuple[1].path)]
+
+        for llo, rlo in upload_files:
+            try:
+                self.dbfs_client.put_file(llo.path, rlo.path, False)
+            except Exception as e:
+                raise RuntimeError('Error \'{}\' while uploading {}'.format(e, llo.path))
+
+        return remote_lib_objects
+
+    @staticmethod
+    def _get_hashed_path(path):
+        """
+        Finds the corresponding dbfs file path for the file located at the supplied path by
+        calculating its hash.
+        :param path: Local File Path
+        :return: Remote Path (pipeline_base_dir + file_hash (dot) file_extension)
+        """
+        hash_buffer = sha1()
+        try:
+            with open(path, 'rb') as f:
+                while True:
+                    data = f.read(BUFFER_SIZE)
+                    if not data:
+                        break
+                    hash_buffer.update(data)
+        except Exception as e:
+            raise RuntimeError('Error \'{}\' while processing {}'.format(e, path))
+
+        file_hash = hash_buffer.hexdigest()
+        path = '{}{}{}'.format(base_pipelines_dir, file_hash, os.path.splitext(path)[1])
```

Comment:
typically, these look better when `base_pipelines_dir` doesn't end with a slash. Also are you missing a period here?

Reply:
I take back the period question. Could you leave a small comment that says splitext includes the period though?